### ISSUE TITLE
[BUG] AuthenticationFailedException: ManagedIdentityCredential authentication failed. An unexpected error occured while fetching the AAD Token

### ISSUE THREAD

**Issue by AnnaGolosova** [Azure/azure-sdk-for-net#40848]
### Library name and version

Azure.Identity 1.10.4

### Describe the bug

Hi all! I'm trying to setup API-to-API authorization using Managed identities. 

I have two web APIs, let's say Client.API and Server.API. In Azure AD Managed Identity service principal for Client.API has appropriate role assigned for Server.API. 

I'm using the next code snippet in Client.API (.NET 6) to get an access token and perform request from Client.API to Server.API:
```
var credentials = new DefaultAzureCredential(new DefaultAzureCredentialOptions
{
    ExcludeSharedTokenCacheCredential = true,
    ExcludeVisualStudioCodeCredential = true,
    ExcludeVisualStudioCredential = true,
    VisualStudioTenantId = "<my-tenant-id>",
});

var tokenResult = credentials.GetToken(new Azure.Core.TokenRequestContext(new[] { "<server-api-app-id>/.default" }));

```

And it failes on the last line with the next error:
`[AuthenticationFailedException] ManagedIdentityCredential authentication failed: Service request failed.\r\nStatus: 500 (Internal Server Error)\r\n\r\nContent:\r\n{\"statusCode\":500,\"message\":\"An unexpected error occured while fetching the AAD Token.\",\"correlationId\":\"d61c989c-ca38-4ab1-aad1-fc1443b89887\"}\r\n\r\nHeaders:\r\nDate: Tue, 19 Dec 2023 11:48:16 GMT\r\nServer: Kestrel\r\nTransfer-Encoding: chunked\r\nX-CORRELATION-ID: REDACTED\r\nContent-Type: application/json; charset=utf-8\r\n\nSee the troubleshooting guide for more information. https://aka.ms/azsdk/net/identity/managedidentitycredential/troubleshoot"`

Stack stace:

### Expected behavior

Authentication through Azure AD Managed Identity works well

### Actual behavior

The next exception was thrown:

```
"exception": {
        "ClassName": "Azure.Identity.AuthenticationFailedException",
        "Message": "ManagedIdentityCredential authentication failed: Service request failed.\r\nStatus: 500 (Internal Server Error)\r\n\r\nContent:\r\n{\"statusCode\":500,\"message\":\"An unexpected error occured while fetching the AAD Token.\",\"correlationId\":\"d61c989c-ca38-4ab1-aad1-fc1443b89887\"}\r\n\r\nHeaders:\r\nDate: Tue, 19 Dec 2023 11:48:16 GMT\r\nServer: Kestrel\r\nTransfer-Encoding: chunked\r\nX-CORRELATION-ID: REDACTED\r\nContent-Type: application/json; charset=utf-8\r\n\nSee the troubleshooting guide for more information. https://aka.ms/azsdk/net/identity/managedidentitycredential/troubleshoot",
        "Data": null,
        "InnerException": {
            "Status": 500,
            "ErrorCode": null,
            "ClassName": "Azure.RequestFailedException",
            "Message": "Service request failed.\r\nStatus: 500 (Internal Server Error)\r\n\r\nContent:\r\n{\"statusCode\":500,\"message\":\"An unexpected error occured while fetching the AAD Token.\",\"correlationId\":\"d61c989c-ca38-4ab1-aad1-fc1443b89887\"}\r\n\r\nHeaders:\r\nDate: Tue, 19 Dec 2023 11:48:16 GMT\r\nServer: Kestrel\r\nTransfer-Encoding: chunked\r\nX-CORRELATION-ID: REDACTED\r\nContent-Type: application/json; charset=utf-8\r\n",
            "Data": null,
            "InnerException": null,
            "HelpURL": null,
            "StackTraceString": "   at Azure.Identity.ManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)\r\n   at Azure.Identity.ManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)\r\n   at Azure.Identity.ManagedIdentityClient.AuthenticateCoreAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)\r\n   at Azure.Identity.ManagedIdentityClient.AppTokenProviderImpl(AppTokenProviderParameters parameters)\r\n   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.SendTokenRequestToAppTokenProviderAsync(ILoggerAdapter logger, CancellationToken cancellationToken)\r\n   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.GetAccessTokenAsync(CancellationToken cancellationToken, ILoggerAdapter logger)\r\n   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.ExecuteAsync(CancellationToken cancellationToken)\r\n   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)\r\n   at Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenForClientParameters clientParameters, CancellationToken cancellationToken)\r\n   at Azure.Identity.AbstractAcquireTokenParameterBuilderExtensions.ExecuteAsync[T](AbstractAcquireTokenParameterBuilder`1 builder, Boolean async, CancellationToken cancellationToken)\r\n   at Azure.Identity.MsalConfidentialClient.AcquireTokenForClientCoreAsync(String[] scopes, String tenantId, Boolean enableCae, Boolean async, CancellationToken cancellationToken)\r\n   at Azure.Identity.MsalConfidentialClient.AcquireTokenForClientAsync(String[] scopes, String tenantId, Boolean enableCae, Boolean async, CancellationToken cancellationToken)\r\n   at Azure.Identity.ManagedIdentityClient.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)\r\n   at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)",
            "RemoteStackTraceString": null,
            "RemoteStackIndex": 0,
            "ExceptionMethod": null,
            "HResult": -2146233088,
            "Source": "Azure.Identity",
            "WatsonBuckets": null
        },
        "HelpURL": null,
        "StackTraceString": "   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex, String additionalMessage, Boolean isCredentialUnavailable)\r\n   at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)\r\n   at Azure.Core.Pipeline.TaskExtensions.EnsureCompleted[T](ValueTask`1 task)\r\n   at Azure.Identity.ManagedIdentityCredential.GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)\r\n   at Azure.Identity.DefaultAzureCredential.GetTokenFromSourcesAsync(TokenCredential[] sources, TokenRequestContext requestContext, Boolean async, CancellationToken cancellationToken)\r\n   at Azure.Identity.DefaultAzureCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)\r\n   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex, String additionalMessage, Boolean isCredentialUnavailable)\r\n   at Azure.Identity.DefaultAzureCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)\r\n   at Azure.Core.Pipeline.TaskExtensions.EnsureCompleted[T](ValueTask`1 task)\r\n   at Azure.Identity.DefaultAzureCredential.GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)",
        "RemoteStackTraceString": null,
        "RemoteStackIndex": 0,
        "ExceptionMethod": null,
        "HResult": -2146233088,
        "Source": "Azure.Identity",
        "WatsonBuckets": null
    }
```

### Reproduction Steps

Create two APIs
Grand a role for managed identity service principal to give an access. (In my case I used New-MgServicePrincipalAppRoleAssignment command in PowerShell)
Try to use `DefaultAzureCredential` class from Azure.Identity nuget package to get an access token via its `GetToken` method.

### Environment

Azure App service running on Windows, .NET 6

**Comment by jsquire** [comment]
Hi @AnnaGolosova.  Thank you for reaching out and we regret that you're experiencing difficulties.  The error that you're seeing indicates that the managed identity endpoint returned a service error to the client.  The `ManagedIdentityCredential` has no insight nor influence over the service behavior.   In this case, the response code and error message indicate that the service itself experienced an exception.

Unfortunately, other than the general guidance in the [Troubleshoot ManagedIdentityCredential authentication issues](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/TROUBLESHOOTING.md#troubleshoot-managedidentitycredential-authentication-issues) section of the Troubleshooting Guide, there's no assistance that we can provide.  Your best path forward would be to [open an Azure support request](https://docs.microsoft.com/azure/azure-portal/supportability/how-to-create-azure-support-request).

I'm going to close this out;  if I've misunderstood what you're describing, please let us know in a comment and we'd be happy to assist as we're able.

**Comment by AnnaGolosova** [comment]
During more deep investigation I've found that I configured access to the wrong target app that caused an issue. So when you get 500 status code it may mean permissions issue (Why not 403 - good question to Azure 😁)
Here's a few steps that helped me to investigate an issue. Open Kudu for Client.API app service. Go to DebugConsole -> CMD and run some of these commands:

`curl -v -H X-IDENTITY-HEADER:%IDENTITY_HEADER% "%IDENTITY_ENDPOINT%?api-version=2019-08-01&resource=https%3A%2F%[2Fvault.azure.net](http://2fvault.azure.net/)"` to check and access to KeyVault
`curl -v -H X-IDENTITY-HEADER:%IDENTITY_HEADER% "%IDENTITY_ENDPOINT%?api-version=2019-08-01&resource=<target-app-reg-id>/.default"` to check an access to the target resource

If all permissions configured fine, you will get 200 (OK) with access token. And in my case 500 was because of missing permissions.
